File: Sources/Deleter.swift
import Foundation

enum Deleter {
    struct Result {
        var deleted: [URL] = []
        var failed: [(url: URL, error: String)] = []
    }

    /// Delete directories sequentially to avoid I/O spikes.
    static func delete(
        urls: [URL],
        onProgress: (Int, Int, URL) -> Void
    ) -> Result {
        var result = Result()
        let fm = FileManager.default

        for (index, url) in urls.enumerated() {
            onProgress(index + 1, urls.count, url)

            do {
                try fm.removeItem(at: url)
                result.deleted.append(url)
            } catch {
                result.failed.append((url: url, error: error.localizedDescription))
            }
        }

        return result
    }
}

File: cli/lib/deleter.js
import fs from 'node:fs/promises';
import path from 'node:path';

/**
 * Delete node_modules directories sequentially with progress reporting.
 * Sequential to avoid I/O spikes from parallel large deletions.
 *
 * @param {string[]} paths - Absolute paths to delete
 * @param {object} options
 * @param {function} [options.onProgress] - Called with (current, total, path)
 * @returns {Promise<{deleted: string[], failed: Array<{path: string, error: string}>}>}
 */
export async function deleteDirectories(paths, options = {}) {
  const { onProgress } = options;
  const deleted = [];
  const failed = [];

  for (let i = 0; i < paths.length; i++) {
    const dirPath = paths[i];

    if (path.basename(dirPath) !== 'node_modules') {
      failed.push({ path: dirPath, error: 'Refusing to delete: path does not end in node_modules' });
      continue;
    }

    if (onProgress) {
      onProgress(i + 1, paths.length, dirPath);
    }

    try {
      await fs.rm(dirPath, { recursive: true, force: true });
      deleted.push(dirPath);
    } catch (err) {
      failed.push({ path: dirPath, error: err.message || 'Unknown error' });
    }
  }

  return { deleted, failed };
}
